File: ponio/examples/nagumo.py
# Copyright 2022 PONIO TEAM. All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

#!/usr/bin/env python

import subprocess
import os
import dataclasses
import glob
import argparse

import matplotlib.pyplot as plt
import numpy as np
import numpy.typing as npt

name = "nagumo"
data_dir = name+"_data"
img_dir = data_dir

parser = argparse.ArgumentParser(
    description=f"Compile, launch and plot results of example `{name}`")
parser.add_argument('--only-save', action='store_true',
                    help="Just save output in img directory")

arguments = parser.parse_args()

if arguments.only_save:
    img_dir = os.path.join("img", name)
    os.makedirs(img_dir, exist_ok=True)

k = 1.0
d = 1.0

x_0 = -10.


@dataclasses.dataclass
class simu_data:
    time: float
    data: npt.NDArray[np.float64]

    def x(self):
        return self.data[:, 0]

    def u(self):
        return self.data[:, 1]

    def exact_sol(self):
        v = (1./np.sqrt(2.))*np.sqrt(k*d)
        cst = -(1./np.sqrt(2.))*np.sqrt(k/d)

        return np.exp(cst*(self.x() - x_0 - v*self.time)) / (1. + np.exp(cst*(self.x() - x_0 - v*self.time)))


def load_data(filename):
    time = float(filename.split("/")[-1][:-4].split("_")[1])
    return simu_data(time=time, data=np.loadtxt(filename))


os.makedirs(data_dir, exist_ok=True)

make = subprocess.Popen(["make", name])
make.wait()

args = [os.path.join(".", name)]
process = subprocess.Popen(args)
process.wait()

data = sorted([load_data(file) for file in glob.glob(
    os.path.join(data_dir, "*.dat"))], key=lambda d: d.time)

for i, dat in enumerate(data):
    plt.plot(dat.x(), dat.u(), color=f"C{i}", label=f"$t^n = {dat.time}$")

plt.legend()

plt.savefig(os.path.join(img_dir, "01.png"))

if not arguments.only_save:
    plt.title("Nagumo solution")
    plt.show()

for i, dat in enumerate(data):
    plt.plot(dat.x(), dat.u() - dat.exact_sol(),
             color=f"C{i}", label=f"$t^n = {dat.time}$")

plt.legend()

plt.savefig(os.path.join(img_dir, "02.png"))

if not arguments.only_save:
    plt.title("Absolute error in time")
    plt.show()

File: ponio/examples/nagumo.cpp
// Copyright 2022 PONIO TEAM. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <iterator>
#include <numbers>
#include <sstream>
#include <string>
#include <valarray>

#include <ponio/runge_kutta.hpp>
#include <ponio/solver.hpp>
#include <ponio/time_span.hpp>

struct nagumo
{
    double dx;
    double x_min;
    double x_max;
    double x_0;
    double k;
    double d;

    nagumo( double ndx, double nx_min, double nx_max, double nx_0, double nk, double nd )
        : dx( ndx )
        , x_min( nx_min )
        , x_max( nx_max )
        , x_0( nx_0 )
        , k( nk )
        , d( nd )
    {
    }

    std::valarray<double>
    operator()( double, std::valarray<double> const& u ) const
    {
        double const oneoverdxdx = 1. / ( dx * dx );
        std::size_t const nx     = u.size();

        std::valarray<double> udot( nx );

        udot[0] = oneoverdxdx * ( -2. * u[0] + 2. * u[1] );
        for ( std::size_t i = 1; i < nx - 1; ++i )
        {
            udot[i] = oneoverdxdx * ( u[i - 1] - 2. * u[i] + u[i + 1] );
        }
        udot[nx - 1] = oneoverdxdx * ( 2. * u[nx - 2] - 2. * u[nx - 1] );

        udot += k * u * u * ( 1.0 - u );

        return udot;
    }

    std::valarray<double>
    exact_solution( double t, std::valarray<double> const& x ) const
    {
        double const v   = ( 1. / std::numbers::sqrt2 ) * ( std::sqrt( k * d ) );
        double const cst = -( 1. / std::numbers::sqrt2 ) * std::sqrt( k / d );

        return std::exp( cst * ( x - x_0 - v * t ) ) / ( 1.0 + std::exp( cst * ( x - x_0 - v * t ) ) );
    }
};

void
save( std::valarray<double> const& x, std::valarray<double> const& y, std::filesystem::path const& dir, double t )
{
    auto printer_xy = []( double a, double b )
    {
        std::stringstream ss;
        ss << a << " " << b;
        return ss.str();
    };

    std::stringstream filename;
    filename << "u_" << std::setprecision( 3 ) << t << ".dat";

    std::ofstream of( dir / filename.str() );
    std::transform( std::begin( x ), std::end( x ), std::begin( y ), std::ostream_iterator<std::string>( of, "\n" ), printer_xy );
    of.close();
}

int
main()
{
    std::string const dirname = "nagumo_data";
    std::filesystem::create_directories( dirname );

    std::size_t const nx = 501;

    double const x_max = 50.0;
    double const x_min = -50.0;
    double const x_0   = -10.0;
    double const k     = 1.0;
    double const d     = 1.0;

    double const dx = ( x_max - x_min ) / static_cast<double>( nx - 1 );

    std::valarray<double> x( nx );
    std::generate( std::begin( x ),
        std::end( x ),
        [x_min, dx, count = 0]() mutable
        {
            return x_min + dx * ( count++ );
        } );

    auto pb = nagumo( dx, x_min, x_max, x_0, k, d );

    double const t_ini = 0.0;
    double const t_end = 50.0;
    double const dt    = ( t_end - t_ini ) / 100;

    ponio::time_span<double> const t_span = { t_ini, t_end };

    std::valarray<double> const u_n = pb.exact_solution( t_ini, x );

    auto sol_range = ponio::make_solver_range( pb, ponio::runge_kutta::rkc_202(), u_n, t_span, dt );
    auto it_sol    = sol_range.begin();

    std::size_t n_iteration = 0;
    while ( it_sol->time < t_end )
    {
        if ( n_iteration % 10 == 0 )
        {
            save( x, it_sol->state, std::filesystem::path( dirname ), it_sol->time );
        }

        ++n_iteration;
        ++it_sol;
    }
    save( x, it_sol->state, std::filesystem::path( dirname ), it_sol->time );

    return 0;
}
